pkg/scale/decode.go

```diff
@@ -0,0 +1,539 @@
+			switch t.ConvertibleTo(reflect.TypeOf(VaryingDataType{})) {
+			case true:
+				err = ds.decodeVaryingDataType(dstv)
+			case false:
+				err = ds.decodeSlice(dstv)
+			}
+		default:
+			_, ok := in.(VaryingDataTypeValue)
+			switch ok {
+			case true:
+				var temp reflect.Value
+				t := reflect.TypeOf(in)
+				switch t.Kind() {
+				// TODO: support more primitive types.  Do we need to support arrays and slices as well?
+				case reflect.Int:
+					temp = reflect.New(reflect.TypeOf(int(1)))
+					err = ds.unmarshal(temp.Elem())
+					if err != nil {
+						break
+					}
+				case reflect.Int16:
+					temp = reflect.New(reflect.TypeOf(int16(1)))
+					err = ds.unmarshal(temp.Elem())
+					if err != nil {
+						break
+					}
+				default:
+					err = fmt.Errorf("unsupported kind for VaryingDataTypeValue: %s", t.Kind())
+					return
+				}
+				dstv.Set(temp.Elem().Convert(t))
+			default:
+				err = fmt.Errorf("unsupported type: %T", in)
+			}
+		}
+	}
+	return
+}
+
+func (ds *decodeState) decodeResult(dstv reflect.Value) (err error) {
+	dstt := reflect.TypeOf(dstv.Interface())
+	key := fmt.Sprintf("%s.%s", dstt.PkgPath(), dstt.Name())
+	resultCases, ok := resCache[key]
+	if !ok {
+		err = fmt.Errorf("unable to find registered custom Result: %T", dstv.Interface())
+		return
+	}
+
+	var rb byte
+	rb, err = ds.ReadByte()
+	if err != nil {
+		return
+	}
+	switch rb {
+	case 0x00:
+		// Ok case
+		okIn, ok := resultCases[true]
+		if !ok {
+			err = fmt.Errorf("unable to find registered custom Result.Ok for: %T", dstv.Interface())
+			return
+		}
+		newOk := reflect.New(reflect.TypeOf(okIn))
+		err = ds.unmarshal(newOk.Elem())
+		if err != nil {
+			break
+		}
+		res := reflect.New(dstt)
+		res.Elem().FieldByName("Ok").Set(newOk.Elem())
+		dstv.Set(res.Elem())
+	case 0x01:
+		// Error case
+		errIn, ok := resultCases[false]
+		if !ok {
+			err = fmt.Errorf("unable to find registered custom Result.Err for: %T", dstv.Interface())
+			return
+		}
+		newErr := reflect.New(reflect.TypeOf(errIn))
+		err = ds.unmarshal(newErr.Elem())
+		if err != nil {
+			break
+		}
+		res := reflect.New(dstt)
+		res.Elem().FieldByName("Err").Set(newErr.Elem())
+		dstv.Set(res.Elem())
+	default:
+		err = fmt.Errorf("unsupported Result value: %v, bytes: %v", rb, ds.Bytes())
+	}
+	return
+}
+
+func (ds *decodeState) decodePointer(dstv reflect.Value) (err error) {
+	var rb byte
+	rb, err = ds.ReadByte()
+	if err != nil {
+		return
+	}
+	switch rb {
+	case 0x00:
+		// nil case
+	case 0x01:
+		elemType := reflect.TypeOf(dstv.Interface()).Elem()
+		tempElem := reflect.New(elemType)
+		err = ds.unmarshal(tempElem.Elem())
+		if err != nil {
+			break
+		}
+		dstv.Set(tempElem)
+	default:
+		err = fmt.Errorf("unsupported Option value: %v, bytes: %v", rb, ds.Bytes())
+	}
+	return
+}
+
+func (ds *decodeState) decodeVaryingDataType(dstv reflect.Value) (err error) {
+	l, err := ds.decodeLength()
```

Comment:
does this work for singular varying data types? like for example, a single `DigestItem` and not a `Digests{}` slice?

Reply:
It would work yes, but it still needs to be treated like a `VaryingDataType` which is a `[]VaryingDataTypeValue`.  In this case it would just have a length of 1.